text/0017-message/0017-message.md

```diff
@@ -0,0 +1,181 @@
+                    </tr>
+                    <tr>
+                        <td>Payload Type</td>
+                        <td>varint</td>
+                        <td>
+                            The type of the payload. It will instruct the node how to parse the fields that follow. Types in the range of 0-127 are "core types" that all nodes are expected to know.
+                        </td>
+                    </tr>
+                    <tr>
+                        <td>Data Fields</td>
+                        <td>ANY</td>
+                        <td>A sequence of fields, where the structure depends on <code>payload type</code>.</td>
+                    </tr>
+                    <tr>
+                        <td>Nested Payload (Optional)</td>
+                        <td> Payload</td>
+                        <td> Some payloads support nesting other payloads inside of them.</td>
+                    </tr>
+                </table>
+            </details>
+    <tr>
+        <td>Nonce</td>
+        <td>uint64</td>
+        <td>The nonce which lets this message fulfill the Proof-of-Work requirement.</td>
+    </tr>
+</table>
+
+### Message PoW Hash
+
+This hash is computed by converting all the messages raw bytes into trytes with `b1t6`. Then hash the result with Curl-P-81. The result will be used to verify the proof-of-work done on the message. This will be used for spam protection.
+
+Ideally, the Message ID would also be used for the PoW, but then the Hash function used for the Message ID must also be used for the PoW. Separating them gives us much more flexibility with the PoW method and we can even apply the current Curl based PoW mechanism.
+
+### Message Validation
+
+A message is considered valid, if the following syntactic rules are met:
+
+1. The message length must not exceed X [tbd] bytes.
+2. Varint fields must adhere to the following rules:
+    * *Malleability protection:* No `0x80` or `0x00` trailing bytes. Take for example the following 3 byte varint: `0x8F8000`. 
+    In order to parse it we look at its bits: 
+    `1000 1111 1000 0000 0000 0000`.
+    First we strip the msbs:
+    `000 1111 000 0000 000 0000`
+    Then we reverse the groups of 7 bits (Little Endian):
+    `000 0000 000 0000 000 1111`
+    Then we concatenate all the bits to finally get:
+    `1111`
+    So the varint `0x0F` equals to `0x8F8000`, but is encoded with different bytes.
+    * *Size Protection*: The varint must not be more than 10 bytes long. A 10 byte varint can encompass all the range of `uint64`. In case of an overflow above the `uint64` range the message is invalid.
+2. When we are done parsing the message there shouldn't be any trailing bytes left that were not parsed.
+4. If the `payload type` is in the core payload range (0-127) and the node is familiar with it, or if it is in the range of 128-16383. The upper limit of 16383 is the maximal value of a 2 byte varint, and it should suffice for all the optional payload types we need.
```

Comment:
How do register a custom (community) payload type?

Is 16383 sufficient for all applications?

Reply:
IIRC, in our meetings we threw ideas on how it might be.
I think one idea was that community member will also create RFCs but not sure this is the best way to approach this

I think that together with community volunteers we should define at least one community payload. Then we should properly document and implement it, so other community members can easily follow.
We should probably talk about it more in a meeting.

Regarding the 16,383 (2^14 - 1) limit, I just thought it was a high enough number...
Given that we support colored coins it is hard for me to imagine that the payload feature will become so popular that this limit won't suffice...

Reply:
We should really explain how we envision this community-driven payload draft process.
I guess we want to have a table somewhere where people can create a PR with a new ID and link to the document / RFC describing the payload.

Does it maybe make more sense to have a separate RFC with this table (similar to [SLIP-44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) or [SLIP-173](https://github.com/satoshilabs/slips/blob/master/slip-0173.md))? That could make the process considerably easier.

Reply:
Why use a varint if you have a max bound of 2 bytes? It is way easier to define a uint16 here

Reply:
See my other answer

Also Wolfgang seems to not like this bound, so maybe this bound will be removed (still need to discuss this with him)

Reply:
I think the bound (for the non-core payloads) seems arbitrary and should be removed. We should allow for the full uint32 maybe even uint64 space.

Reply:
I agree the message length protection suffices.
Actually, while thinking about the case for the bound, I started to question whether the message length is really needed...